katdal/schemas/__init__.py

```diff
@@ -0,0 +1,58 @@
+"""Makes packaged XSD schemas available as validators."""
+
+import pkg_resources
+
+
+has_lxml = False
+validators = {}
+
+try:
+    from lxml import etree
+    has_lxml = True
+except ImportError:
+    pass
+
+
+class ValidatorWithLog(object):
+    def __init__(self, validator):
+        self.validator = validator
+
+    def validate(self, xml_string):
+        """Validates a supplied XML string against the instantiated validator.
+
+        Parameters
+        ---------
+        xml_string : str
+            String representation of the XML to be turned into a document
+            and validated.
+
+        Raises
+        ------
+        etree.DocumentInvalid
+            if `xml_string` does not validate against the XSD schema
+        ValueError
+            if `xml_string` cannot be parsed into a valid XML document
+        """
+        try:
+            xml_doc = etree.fromstring(bytes(bytearray(xml_string, encoding='utf-8')))
+        except etree.XMLSyntaxError as e:
+            raise ValueError(e)
+        if not self.validator.validate(xml_doc):
+            log = self.validator.error_log
+            raise etree.DocumentInvalid(log.last_error)
```

Comment:
You are converting `etree.XMLSyntaxError` to `ValueError` but leaving `etree.DocumentInvalid` as is... This seems a bit inconsistent. Will the end user typically be expected to catch invalid document exceptions? If that's the case, I'd probably also transmogrify it for the user's convenience.

Reply:
Simply because the XMLSyntaxError is quite messy. I also wanted to easily catch the difference between an XML syntax error and an invalid document. If I didn't use DocumentInvalid, then ValueError would be the sensible standard exception but wouldn't allow this differentiation.